### ISSUE TITLE
DeadLetter when using Cluster client

### ISSUE THREAD

**Issue by hepin1989** [akka/akka#17977]
server side play,client side akka,version:akka2.4 -m2
using the ClusterClientReceptionist,it didnot work

server side :
```
private val receptionist = ClusterClientReceptionist(context.system)
 receptionist.registerService(self)
//the self's address is akka://application/user/listeners/default
```
conf:
```
  akka.cluster.client.receptionist.name = "payservice"
```
the client side :
```
      receiptClient ? Send(s"/user/listeners/$name", UUID.randomUUID().toString, localAffinity = true)
//name is "default"
```
and the conf:
```
    initial-contacts = ["akka.tcp://application@127.0.0.1:2552/system/payservice"]

```
I don't know why this error came out,but I could make it works by creating the``` ClusterReceptionist ```by hand.

```
[INFO] [07/13/2015 18:51:27.712] [application-akka.actor.default-dispatcher-2] [akka://application/system/payservice] Message [akka.cluster.client.ClusterReceptionist$Internal$GetContacts$] from Actor[akka.tcp://gear-runtime@127.0.0.1:2553/user/$a#-179211648] to Actor[akka://application/system/payservice] was not delivered. [6] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.

```

**Comment by hepin1989** [comment]
btw the document section 
```
The initialContacts parameter is a Set[ActorSelection], which can be created like this:

val initialContacts = Set(
  system.actorSelection("akka.tcp://OtherSys@host1:2552/user/receptionist"),
  system.actorSelection("akka.tcp://OtherSys@host2:2552/user/receptionist"))
```
should s/user/system cause the default receptionist is a system actor when you are using it via Extension.

**Comment by patriknw** [comment]
@hepin1989 I think this has been fixed in 2.4.0-RC1, can you confirm?
http://doc.akka.io/docs/akka/2.4.0-RC1/scala/cluster-client.html

**Comment by hepin1989** [comment]
I'm on it

**Comment by hepin1989** [comment]
it broken too.
- my old code which create the receptionist by hands working,it was running with play 2.4,and after upgrade to akka2.4-m1,it can't receive any message.
```
[Error][2015-08-26 20:55:05]: dropping message [class akka.actor.ActorSelectionMessage] for non-local recipient [Actor[akka.tcp://application@127.0.0.1:2552/]] arriving at [akka.tcp://application@127.0.0.1:2552] inbound addresses are [akka.tcp://gear-runtime@127.0.0.1:2552]

   --> : akka.tcp://gear-runtime@127.0.0.1:2552/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2Fgear-runtime%40127.0.0.1%3A2552-1/endpointWriter
```
here the Play server with application akka system name :```application```,and the akka application with the name of :```gear-runtime```.I am let the play application as a cluster,and let the akka application to use the ```cluster-client```.

after I change akka application's system name to ```application```,it still don't work.
```
[Debug][2015-08-26 21:01:05]: Receptionist not available, buffering message type [akka.cluster.pubsub.DistributedPubSubMediator$Send]
   --> : akka.tcp://application@127.0.0.1:2552/user/$a
[Debug][2015-08-26 21:01:05]: order notify actor for :[NewOrder(e36f171b-8c9d-435d-a51d-0416bfa26298,1440594065909,1,1.0,1)] pre start.
   --> : akka.tcp://application@127.0.0.1:2552/user/$Ac
[Debug][2015-08-26 21:01:07]: Sending GetContacts to [ActorSelection[Anchor(akka://application/), Path(/user/payservice)],ActorSelection[Anchor(akka://application/), Path(/user/payservice)]]
   --> : akka.tcp://application@127.0.0.1:2552/user/$a
```
- I try to use the standed version,which use the extention directly ,and still the same,no dead letter,but can't receive any message anymore.


in short.the probleam is not fixed and my code is broken :).and another new probleam is why the new version need the cluster client 's akka system should have the same name as which it is going to connected to ?I remember that it didn't need to be the same.

**Comment by patriknw** [comment]
Thanks for checking. Can you show how exactly you create the client and the receptionist?

    Sending GetContacts to [ActorSelection[Anchor(akka://application/), Path(/user/payservice)]

Why `akka://application/` ? Looks like no remote protocol enabled there?

I would like to see how you define the `initialContacts`.
I would guess that you have defined it as "/user/payservice" instead of using full address "akka.tcp://gear-runtime@127.0.0.1:2552/user/payservice"

**Comment by hepin1989** [comment]
```
#without this ,it will cause dropping,default value is gear-runtime
engine.system-name="application"

engine.plugins {
  pay-client{
    target-game ="default"
    # Actor paths of the ClusterReceptionist actors on the servers (cluster nodes)
    # that the client will try to contact initially. It is mandatory to specify
    # at least one initial contact.
    # Comma separated full actor paths defined by a string on the form of
    # "akka.tcp://system@hostname:port/system/receptionist"
    initial-contacts = ["akka.tcp://application@127.0.0.1:2552/user/payservice"]

    # Interval at which the client retries to establish contact with one of
    # ClusterReceptionist on the servers (cluster nodes)
    establishing-get-contacts-interval = 3s

    # Interval at which the client will ask the ClusterReceptionist for
    # new contact points to be used for next reconnect.
    refresh-contacts-interval = 60s

    # How often failure detection heartbeat messages should be sent
    heartbeat-interval = 2s

    # Number of potentially lost/delayed heartbeats that will be
    # accepted before considering it to be an anomaly.
    # The ClusterClient is using the akka.remote.DeadlineFailureDetector, which
    # will trigger if there are no heartbeats within the duration
    # heartbeat-interval + acceptable-heartbeat-pause, i.e. 15 seconds with
    # the default settings.
    acceptable-heartbeat-pause = 13s

    # If connection to the receptionist is not established the client will buffer
    # this number of messages and deliver them the connection is established.
    # When the buffer is full old messages will be dropped when new messages are sent
    # via the client. Use 0 to disable buffering, i.e. messages will be dropped
    # immediately if the location of the singleton is unknown.
    # Maximum allowed buffer size is 10000.
    buffer-size = 1000
  }
  enabled += "qgame.engine.plugins.PayClientPlugin"
}

engine.components{
  enabled = ["sample.PayClientSampleComponent"]
  pay-client-sample{
    modules{
      enabled = ["sample-module"]
    }
  }
}

engine.akka{
  io-netty.tcp{
    port = 2553
    hostname = "127.0.0.1"
  }
}
```
and the other node is play application
```
# This is the main configuration file for the application.
# ~~~~~

# Secret key
# ~~~~~
# The secret key is used to secure cryptographics functions.
#
# This must be changed for production, but we recommend not changing it in this file.
#
# See http://www.playframework.com/documentation/latest/ApplicationSecret for more details.
play.crypto.secret = "changeme"

# The application languages
# ~~~~~
play.i18n.langs = [ "en" ]

slick.dbs.default.driver="slick.driver.MySQLDriver$"
slick.dbs.default.db.driver="com.mysql.jdbc.Driver"
slick.dbs.default.db.url="jdbc:mysql://127.0.0.1:3306/pay?characterEncoding=utf8"
slick.dbs.default.db.user="root"
slick.dbs.default.db.password="root"

# Evolutions
# ~~~~~
# You can disable evolutions if needed
play.evolutions.enabled=false

# You can disable evolutions for a specific datasource if necessary
play.evolutions.db.default.enabled=false
play.akka{
  loglevel = "DEBUG"
  loggers = ["qgame.engine.logging.ConsoleLogger","akka.event.slf4j.Slf4jLogger"]
  actor{
    provider = "akka.cluster.ClusterActorRefProvider"
  }
}

akka.actor.provider = "akka.cluster.ClusterActorRefProvider"

play{
  akka.extensions = ["akka.cluster.client.ClusterClientReceptionist"]
  # //#receptionist-ext-config
  # Settings for the ClusterClientReceptionist extension
  akka.remote {
    enabled-transports = ["akka.remote.netty.tcp"]
    netty.tcp {
      hostname = "127.0.0.1"
      port = 2552
    }
  }
  akka.cluster.client.receptionist {
    # Actor name of the ClusterReceptionist actor, /system/receptionist
    name = "payservice"

    # Start the receptionist on members tagged with this role.
    # All members are used if undefined or empty.
    role = ""

    # The receptionist will send this number of contact points to the client
    number-of-contacts = 3

    # The actor that tunnel response messages to the client will be stopped
    # after this time of inactivity.
    response-tunnel-receive-timeout = 30s

    # The id of the dispatcher to use for ClusterReceptionist actors.
    # If not specified default dispatcher is used.
    # If specified you need to define the settings of the actual dispatcher.
    use-dispatcher = ""
  }
}

play.cache.bindCaches = ["config-cache"]
```
and the code on the client side is :
```scala
class PayClientAPIImpl(
      system: ActorSystem,
      name: String,
      config: Config
  ) extends PayClientAPI {
    private val requestPath = s"/user/listeners/$name"
    private val settings = ClusterClientSettings(config)

    private val receiptClient = system.actorOf(ClusterClient.props(settings))

    override def order(newOrder: NewOrder)(implicit timeout: Timeout): Future[UnPayedOrder] = {
      val promise = Promise[UnPayedOrder]()
      system.actorOf(OrderNotifyActor.props(newOrder, timeout, promise))
      promise.future
    }

    override def registerGlobalListener(actorRef: ActorRef): Unit = {
      system.actorOf(OrderGlobalNotificationSubscriber.props(OrderGlobalNotificationSubscriber.ActorSubscriber(actorRef)))
    }

    override def registerOnGlobalNotification(thunk: (OrderNotify) => Unit): Unit = {
      system.actorOf(OrderGlobalNotificationSubscriber.props(OrderGlobalNotificationSubscriber.ThunkSubscriber(thunk)))
    }

    override def registerOnGlobalNotification(thunk: Consumer[OrderNotify]): Unit = {
      system.actorOf(OrderGlobalNotificationSubscriber.props(OrderGlobalNotificationSubscriber.ThunkSubscriber(thunk)))
    }

    private class OrderNotifyActor(
        newOrder: NewOrder,
        timeout: Timeout,
        promise: Promise[UnPayedOrder]
    ) extends Actor with ActorLogging {
      //send new order to server
      receiptClient ! Send(requestPath, newOrder, localAffinity = true)

      override def receive: Receive = {
        case unPayedOrderData: UnPayedOrderData =>
          val notifyPromise = Promise[OrderNotify]()
          promise.trySuccess(UnPayedOrder(unPayedOrderData, notifyPromise.future))
          context.become(waitingNotify(sender(), notifyPromise))
      }

      private def waitingNotify(replayTo: ActorRef, notifyPromise: Promise[OrderNotify]): Receive = {
        case orderNotifyData: OrderNotifyData =>
          val ackPromise = Promise[OrderNotifyACK]()
          val replyTo = sender()
          notifyPromise.trySuccess(OrderNotify(orderNotifyData, ackPromise))
          import scala.concurrent.ExecutionContext.Implicits.global
          ackPromise.future.onComplete {
            case Success(v) =>
              self ! OrderNotifyActor.OrderNotifyAckSuccess(v)
            case Failure(e) =>
              self ! OrderNotifyActor.OrderNotifyAckFailed(e)
          }
          context.become(waitingAck(replyTo, orderNotifyData))
      }

      private def waitingAck(replayTo: ActorRef, orderNotifyData: OrderNotifyData): Receive = {
        case OrderNotifyActor.OrderNotifyAckSuccess(ack) =>
          log.debug("ack for id :[{}],status:[{}]", orderNotifyData.id, ack.info)
          replayTo ! ack
          context.stop(self)
        case OrderNotifyActor.OrderNotifyAckFailed(e) =>
          log.error(e, "ack for id :[{}]", orderNotifyData.id)
          context.stop(self)
      }

      @throws[Exception](classOf[Exception])
      override def preStart(): Unit = {
        log.info("order notify actor for :[{}] request path :[{}]", newOrder,requestPath)
        super.preStart()
      }
    }

    private class OrderGlobalNotificationSubscriber(subscriber: OrderGlobalNotificationSubscriber.Subscriber) extends Actor with ActorLogging {
      receiptClient ! Send(requestPath, SubscribeToOrderNotifier, localAffinity = true)

      override def receive: Receive = {
        case Subscribed =>
          log.debug("subscribed to :[{}]", sender())
          context.become(subscribed)

      }

      private val subscribed: Receive = {
        case orderNotifyData: OrderNotifyData =>
          log.debug("received order notification via publish channel")
          val replyTo = sender()
          try {
            val promise = Promise[OrderNotifyACK]()
            subscriber.notify(OrderNotify(orderNotifyData, promise))
            promise.future.onComplete {
              case Success(ack) =>
                self ! OrderGlobalNotificationSubscriber.NotifyOrderNotifySuccess(replyTo, orderNotifyData, ack)
              case Failure(e) =>
                self ! OrderGlobalNotificationSubscriber.NotifyOrderNotifyFailed(replyTo, orderNotifyData, e)
            }
          } catch {
            case NonFatal(e) =>
              log.error(e, "error occur  when notify order notify :[{}]", orderNotifyData)
              self ! OrderGlobalNotificationSubscriber.NotifyOrderNotifyFailed(replyTo, orderNotifyData, e)
          }

        case OrderGlobalNotificationSubscriber.NotifyOrderNotifySuccess(replyTo, orderNotifyData, ack) =>
          log.debug("reply to :[{}] for data :[{}] with ack :[{}]", replyTo, orderNotifyData, ack)
          replyTo ! ack
        case OrderGlobalNotificationSubscriber.NotifyOrderNotifyFailed(replyTo, orderNotifyData, e) =>
          log.error(e, "error when reply to :[{}] for data :[{}] with error :[{}]", replyTo, orderNotifyData, e.getMessage)
          replyTo ! OrderNotifyACK(orderNotifyData.id, System.currentTimeMillis(), success = false, e.getMessage)

      }

    }

    object OrderNotifyActor {
      def props(
        newOrder: NewOrder,
        timeout: Timeout,
        promise: Promise[UnPayedOrder]
      ): Props = Props(new OrderNotifyActor(newOrder, timeout, promise))

      sealed trait OrderNotifyActorEvent

      case class OrderNotifyAckSuccess(ack: OrderNotifyACK) extends OrderNotifyActorEvent

      case class OrderNotifyAckFailed(cause: Throwable) extends OrderNotifyActorEvent

    }

    object OrderGlobalNotificationSubscriber {

      def props(subscriber: Subscriber): Props = Props(new OrderGlobalNotificationSubscriber(subscriber))

      sealed trait Subscriber {
        def notify(orderNotify: OrderNotify): Unit
      }

      sealed trait OrderGlobalNotificationSubscriberEvent

      case class ActorSubscriber(actorRef: ActorRef) extends Subscriber {
        override def notify(orderNotify: OrderNotify): Unit = actorRef ! orderNotify
      }

      case class ThunkSubscriber(thunk: OrderNotify => Unit) extends Subscriber {
        override def notify(orderNotify: OrderNotify): Unit = thunk(orderNotify)
      }

      case class NotifyOrderNotifySuccess(replyTo: ActorRef, orderNotifyData: OrderNotifyData, orderNotifyACK: OrderNotifyACK) extends OrderGlobalNotificationSubscriberEvent

      case class NotifyOrderNotifyFailed(replyTo: ActorRef, orderNotifyData: OrderNotifyData, e: Throwable) extends OrderGlobalNotificationSubscriberEvent

      object ThunkSubscriber {
        def apply(thunk: Consumer[OrderNotify]): ThunkSubscriber = apply(thunk.accept _)
      }

    }

  }
```
and on the play side is
```
@Singleton
class OrderService @Inject() (system: ActorSystem, orderDao: OrderDao) {
  Cluster(system).join(Cluster(system).selfAddress)

  private val listenerManager = system.actorOf(OrderListenerManager.props(orderDao), "listeners")

  def handleReceipt(game: String, channel: String, receipt: Receipt): Future[OrderNotifyACK] = {
    val promise = Promise[OrderNotifyACK]()
    listenerManager ! HandleReceipt(game, channel, receipt, promise)
    promise.future
  }

}

class OrderListenerManager(orderDao: OrderDao) extends Actor with ActorLogging {
  private val mediator = DistributedPubSub(context.system).mediator
  private val config = context.system.settings.config.getConfig("akka.cluster.client.receptionist")
  private val receptionist = context.system.actorOf(ClusterReceptionist.props(mediator, ClusterReceptionistSettings(config)), "payservice")
  private var listenersMapping: Map[String, ActorRef] = Map.empty
  Logger.error(s"receipt address :$receptionist")

  override def receive: Actor.Receive = {
    case CreateListener(name, promise) =>
      listenersMapping.get(name) match {
        case Some(pre) =>
          log.warning("try to create listener for name:{},but exists.", name)
          promise.success(s"listener :[$name] already exists at :[$pre]")
        case None =>
          log.info("try to create listener for name:{}.", name)
          //game name -> actorRef
          val ref = context.actorOf(OrderListener.props(name, orderDao), name)
          listenersMapping = listenersMapping.updated(name, ref)
      }
    case handle @ HandleReceipt(game, channel, receipt, promise) =>
      log.info("start handle receipt for game :[{}],channel:[{}]", game, channel)
      context.child(game) match {
        case Some(ref) =>
          ref forward handle
        case None =>
          promise.failure(new IllegalArgumentException(s"game :[$game] is not supported,no order listener for it."))
      }
  }

  @throws[Exception](classOf[Exception])
  override def preStart(): Unit = {
    super.preStart()
    self ! CreateListener("default", Promise[String]())
  }
}

object OrderListenerManager {
  def props(orderDao: OrderDao) = Props(new OrderListenerManager(orderDao))

  sealed trait OrderListenerManagerCommand

  case class CreateListener(name: String, promise: Promise[String]) extends OrderListenerManagerCommand

  case class HandleReceipt(game: String, channel: String, receipt: Receipt, promise: Promise[OrderNotifyACK]) extends OrderListenerManagerCommand

}

// user/listeners/$name
class OrderListener(
    name: String,
    orderDao: OrderDao
) extends Actor with ActorLogging {
  private val mediator = DistributedPubSub(context.system).mediator
  private val notifier = context.actorOf(OrderNotifier.props(orderDao), "notifier")

  override def receive: Receive = {
    case newOrder: NewOrder =>
      val newOrderWithUUID = NewOrderWithUUID(newOrder)
      context.actorOf(OrderWorker.props(orderDao, newOrderWithUUID, sender()), newOrderWithUUID.id)
    case msg @ SubscribeToOrderNotifier =>
      notifier forward msg
    case handle @ HandleReceipt(game, channel, receipt, promise) =>
      log.info("handle receipt from game:[{}] channel :[{}],receipt :[{}]", game, channel, receipt)
      try {
        val orderNotifyData = receipt.toOrderNotifyData
        val id = orderNotifyData.id
        context.child(id) match {
          case Some(ref) =>
            log.debug("forwarding receipt handling to worker id :[{}],from channel:[{}] receipt :[{}]", id, channel, receipt)
            ref forward handle
          case None =>
            log.warning("could not found worker actor for id :[{}] from channel :[{}],will try notifier", id, channel)
            //TODO this id maybe the previous id?should published to the cluster notification
            //promise.failure(new IllegalArgumentException(s"id :[{}] is not a valid order id ,don't find"))
            notifier ! NotifySubscribersGlobally(game, channel, orderNotifyData, promise)
        }
      } catch {
        case NonFatal(e) =>
          log.error(e, "error when handle order notify data for channel :[{}],receipt :[{}]")
          promise.failure(e)
      }
    case msg =>
      Logger.info(s"msg:[$msg],from:[${sender()}}]")
      sender() ! msg
  }

  @throws[Exception](classOf[Exception])
  override def preStart(): Unit = {
    log.info("listener :[{}] preStart", name)
    log.info("started for listener name :[{}] at :[{}]", name, self.path)
    mediator ! DistributedPubSubMediator.Put(self)
    super.preStart()
  }

  @throws[Exception](classOf[Exception])
  override def postStop(): Unit = {
    log.info("listener :[{}] postStop", name)
    super.postStop()
  }
}
```

I have tried the standand one ,with
```
class OrderListenerManager(orderDao: OrderDao) extends Actor with ActorLogging {
  private val receptionist = ClusterClientReceptionist(context.system)
  private var listenersMapping: Map[String, ActorRef] = Map.empty
  Logger.error(s"receipt address :${receptionist}")

  override def receive: Actor.Receive = {
    case CreateListener(name, promise) =>
      listenersMapping.get(name) match {
        case Some(pre) =>
          log.warning("try to create listener for name:{},but exists.", name)
          promise.success(s"listener :[$name] already exists at :[$pre]")
        case None =>
          log.info("try to create listener for name:{}.", name)
          //game name -> actorRef
          val ref = context.actorOf(OrderListener.props(name, orderDao), name)
          receptionist.registerService(ref)
          listenersMapping = listenersMapping.updated(name, ref)
      }
    case handle @ HandleReceipt(game, channel, receipt, promise) =>
      log.info("start handle receipt for game :[{}],channel:[{}]", game, channel)
      context.child(game) match {
        case Some(ref) =>
          ref forward handle
        case None =>
          promise.failure(new IllegalArgumentException(s"game :[$game] is not supported,no order listener for it."))
      }
  }

  @throws[Exception](classOf[Exception])
  override def preStart(): Unit = {
    super.preStart()
    self ! CreateListener("default", Promise[String]())
  }
}
```
and change the initial-contacts to 
```initial-contacts = ["akka.tcp://application@127.0.0.1:2552/system/receptionist"]```

**Comment by patriknw** [comment]
First, I have verified that the ActorSystem name is not the problem. Evidence:

    [JVM-1] [INFO] [08/26/2015 19:34:38.828] [Sys2-akka.actor.default-dispatcher-15] [akka.trttl.gremlin.tcp://Sys2@localhost:2552/user/client1b] Connected to [akka.trttl.gremlin.tcp://ClusterClientSpec@localhost:58942/system/receptionist]

Then I have tried several things in addition to the tests that we already have and all is working as expected. I can't see anything that is wrong with the code that you supplied, but there is probably just some small mistake somewhere.

When you debug this the first thing you should look at is the client debug log message starting with "Sending GetContacts to". The one you wrote earlier is not going to work because the anchor of the actor selection is not the remote server.

    [Debug][2015-08-26 21:01:07]: Sending GetContacts to [ActorSelection[Anchor(akka://application/), Path(/user/payservice)],ActorSelection[Anchor(akka://application/), Path(/user/payservice)]]
   --> : akka.tcp://application@127.0.0.1:2552/user/$a

**Comment by hepin1989** [comment]
I found the problem.it should be
```
engine.akka{
remote{
      io-netty.tcp{
                  port = 2553
                  hostname = "127.0.0.1"
      }
}
}

```
Sorry for that.

**Comment by patriknw** [comment]
Glad to hear that. No problem

**Comment by hepin1989** [comment]
this is fixed.

**Comment by hepin1989** [comment]
thanks again,the problem why it shows up the ```Anchor(akka://application/)``` is because that the I miss configured the configuration file,so the port is the default one,then 
akka.tcp://application@127.0.0.1:2552/ one becomes akka://application/

**Comment by hepin1989** [comment]
I can confirmed that this is fixed,close it now.
